File: MCP/mysql/db_operations.py
import json
import sys
import os
import pymysql
from pymysql.cursors import DictCursor
from tabulate import tabulate

# 读取配置文件
config_path = os.path.join(os.path.dirname(__file__), 'HieTic_and_SceAr.json')
with open(config_path, 'r') as f:
    config = json.load(f)

def create_connection(db_name):
    """创建数据库连接"""
    connection_info = next((conn for conn in config['connections'] if conn['name'] == db_name), None)
    
    if not connection_info:
        print(f'找不到数据库 "{db_name}" 的配置信息')
        return None
    
    try:
        connection = pymysql.connect(
            host=connection_info['host'],
            port=connection_info['port'],
            user=connection_info['user'],
            password=connection_info['password'],
            database=connection_info['database'],
            cursorclass=DictCursor
        )
        print(f'成功连接到 {db_name} 数据库')
        return connection
    except Exception as e:
        print(f'连接到 {db_name} 数据库时出错: {e}')
        return None

def show_tables(db_name):
    """功能1: 查看数据库中的所有表"""
    connection = create_connection(db_name)
    if not connection:
        return
    
    try:
        with connection.cursor() as cursor:
            cursor.execute('SHOW TABLES')
            tables = cursor.fetchall()
            
            print(f'{db_name} 数据库中的表:')
            for i, table in enumerate(tables, 1):
                # 表名是字典的第一个值
                table_name = list(table.values())[0]
                print(f'{i}. {table_name}')
            
            return tables
    except Exception as e:
        print(f'查询表时出错: {e}')
    finally:
        connection.close()

def describe_table(db_name, table_name):
    """功能2: 查看指定表的结构"""
    connection = create_connection(db_name)
    if not connection:
        return
    
    try:
        with connection.cursor() as cursor:
            cursor.execute(f'DESCRIBE {table_name}')
            structure = cursor.fetchall()
            
            print(f'表 {table_name} 的结构:')
            if structure:
                headers = structure[0].keys()
                table_data = [[row[col] for col in headers] for row in structure]
                print(tabulate(table_data, headers=headers, tablefmt='grid'))
            
            return structure
    except Exception as e:
        print(f'查询表 {table_name} 结构时出错: {e}')
    finally:
        connection.close()

def query_table(db_name, table_name, limit=10):
    """功能3: 查看指定表的数据"""
    connection = create_connection(db_name)
    if not connection:
        return
    
    try:
        with connection.cursor() as cursor:
            cursor.execute(f'SELECT * FROM {table_name} LIMIT {limit}')
            rows = cursor.fetchall()
            
            print(f'表 {table_name} 的数据 (限制 {limit} 条):')
            if rows:
                headers = rows[0].keys()
                table_data = [[row[col] for col in headers] for row in rows]
                print(tabulate(table_data, headers=headers, tablefmt='grid'))
            else:
                print("表中没有数据")
            
            return rows
    except Exception as e:
        print(f'查询表 {table_name} 数据时出错: {e}')
    finally:
        connection.close()

def print_usage():
    """打印使用说明"""
    print('用法:')
    print('python db_operations.py show-tables <数据库名>')
    print('python db_operations.py describe-table <数据库名> <表名>')
    print('python db_operations.py query-table <数据库名> <表名> [限制数量]')

def main():
    """主函数"""
    if len(sys.argv) < 3:
        print_usage()
        return
    
    command = sys.argv[1]
    db_name = sys.argv[2]
    
    if command == 'show-tables':
        show_tables(db_name)
    elif command == 'describe-table':
        if len(sys.argv) < 4:
            print('需要提供表名')
            return
        table_name = sys.argv[3]
        describe_table(db_name, table_name)
    elif command == 'query-table':
        if len(sys.argv) < 4:
            print('需要提供表名')
            return
        table_name = sys.argv[3]
        limit = int(sys.argv[4]) if len(sys.argv) > 4 else 10
        query_table(db_name, table_name, limit)
    else:
        print(f'未知命令: {command}')
        print_usage()

if __name__ == '__main__':
    main() 

File: MCP/mysql/db_operations.js
const fs = require('fs');
const mysql = require('mysql2/promise');
const path = require('path');

// 读取配置文件
const configPath = path.join(__dirname, 'HieTic_and_SceAr.json');
const config = JSON.parse(fs.readFileSync(configPath, 'utf8'));

// 建立数据库连接
async function createConnection(dbName) {
  const connectionInfo = config.connections.find(conn => conn.name === dbName);
  
  if (!connectionInfo) {
    console.error(`找不到数据库 "${dbName}" 的配置信息`);
    return null;
  }
  
  try {
    const connection = await mysql.createConnection({
      host: connectionInfo.host,
      port: connectionInfo.port,
      user: connectionInfo.user,
      password: connectionInfo.password,
      database: connectionInfo.database
    });
    console.log(`成功连接到 ${dbName} 数据库`);
    return connection;
  } catch (error) {
    console.error(`连接到 ${dbName} 数据库时出错:`, error);
    return null;
  }
}

// 功能1: 查看数据库中的所有表
async function showTables(dbName) {
  const connection = await createConnection(dbName);
  if (!connection) return;
  
  try {
    const [rows] = await connection.execute('SHOW TABLES');
    console.log(`${dbName} 数据库中的表:`);
    rows.forEach((row, index) => {
      console.log(`${index + 1}. ${Object.values(row)[0]}`);
    });
    return rows;
  } catch (error) {
    console.error('查询表时出错:', error);
  } finally {
    await connection.end();
  }
}

// 功能2: 查看指定表的结构
async function describeTable(dbName, tableName) {
  const connection = await createConnection(dbName);
  if (!connection) return;
  
  try {
    const [rows] = await connection.execute(`DESCRIBE ${tableName}`);
    console.log(`表 ${tableName} 的结构:`);
    console.table(rows);
    return rows;
  } catch (error) {
    console.error(`查询表 ${tableName} 结构时出错:`, error);
  } finally {
    await connection.end();
  }
}

// 功能3: 查看指定表的数据
async function queryTable(dbName, tableName, limit = 10) {
  const connection = await createConnection(dbName);
  if (!connection) return;
  
  try {
    const [rows] = await connection.execute(`SELECT * FROM ${tableName} LIMIT ${limit}`);
    console.log(`表 ${tableName} 的数据 (限制 ${limit} 条):`);
    console.table(rows);
    return rows;
  } catch (error) {
    console.error(`查询表 ${tableName} 数据时出错:`, error);
  } finally {
    await connection.end();
  }
}

// 命令行参数处理
async function main() {
  const args = process.argv.slice(2);
  const command = args[0];
  const dbName = args[1];
  
  if (!command || !dbName) {
    console.log('用法:');
    console.log('node db_operations.js show-tables <数据库名>');
    console.log('node db_operations.js describe-table <数据库名> <表名>');
    console.log('node db_operations.js query-table <数据库名> <表名> [限制数量]');
    return;
  }
  
  switch (command) {
    case 'show-tables':
      await showTables(dbName);
      break;
    case 'describe-table':
      const tableName1 = args[2];
      if (!tableName1) {
        console.error('需要提供表名');
        return;
      }
      await describeTable(dbName, tableName1);
      break;
    case 'query-table':
      const tableName2 = args[2];
      const limit = parseInt(args[3]) || 10;
      if (!tableName2) {
        console.error('需要提供表名');
        return;
      }
      await queryTable(dbName, tableName2, limit);
      break;
    default:
      console.error('未知命令:', command);
      break;
  }
}

main().catch(console.error); 